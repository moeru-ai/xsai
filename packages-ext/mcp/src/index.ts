export * from './utils/get-tools'
