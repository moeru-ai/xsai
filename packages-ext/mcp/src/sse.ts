import type { SSEClientTransportOptions } from '@modelcontextprotocol/sdk/client/sse.js'

import { Client } from '@modelcontextprotocol/sdk/client/index.js'
import { SSEClientTransport } from '@modelcontextprotocol/sdk/client/sse.js'

export type CreateSSEClientsOptions = Record<string, SSEClientTransportOptions & {
  url: URL
}>

/** @experimental */
export const createSSEClients = async (mcpServers: CreateSSEClientsOptions, version: string = '1.0.0'): Promise<Record<string, Client>> =>
  Object.fromEntries(
    await Promise.all(
      Object.entries(mcpServers)
        .map(async ([serverName, { url, ...options }]) => {
          const transport = new SSEClientTransport(url, options)
          const client = new Client({ name: serverName, version })
          await client.connect(transport)

          return [serverName, client] as const
        }),
    ),
  )
