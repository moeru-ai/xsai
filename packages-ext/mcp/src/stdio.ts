import type { StdioServerParameters } from '@modelcontextprotocol/sdk/client/stdio.js'

import { Client } from '@modelcontextprotocol/sdk/client/index.js'
import { StdioClientTransport } from '@modelcontextprotocol/sdk/client/stdio.js'
import { stderr } from 'node:process'

export type CreateStdioClientsOptions = Record<string, StdioServerParameters>

/** @experimental */
export const createStdioClients = async (mcpServers: CreateStdioClientsOptions, version: string = '1.0.0'): Promise<Record<string, Client>> =>
  Object.fromEntries(
    await Promise.all(
      Object.entries(mcpServers)
        .map(async ([serverName, server]) => {
          const transport = new StdioClientTransport({ ...server, stderr })
          const client = new Client({ name: serverName, version })
          await client.connect(transport)

          return [serverName, client] as const
        }),
    ),
  )
