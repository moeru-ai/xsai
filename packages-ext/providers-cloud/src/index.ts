export * from './providers'

// TODO: providers list
