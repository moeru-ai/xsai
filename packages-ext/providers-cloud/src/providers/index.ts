export { createDeepSeek } from './deepseek'
export { createFireworksAI } from './fireworks-ai'
export { createGoogleGenerativeAI } from './google-generative-ai'
export { createGroq } from './groq'
export { createMinimax } from './minimax'
export { createMistralAI } from './mistral-ai'
export { createMoonshot } from './moonshot'
export { createNovita } from './novita'
export { createOpenAI } from './openai'
export { createOpenRouter } from './openrouter'
export { createPerplexity } from './perplexity'
export { createQwen } from './qwen'
export { createSiliconFlow } from './silicon-flow'
export { createStepfun } from './stepfun'
export { createTencentHunyuan } from './tencent-hunyuan'
export { createTogetherAI } from './together-ai'
export { createWorkersAI } from './workers-ai'
export { createXAI } from './xai'
export { createZhipu } from './zhipu'
