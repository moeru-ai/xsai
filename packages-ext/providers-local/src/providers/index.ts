export { createClaude2OpenAI } from './claude2openai'
export { createOllama } from './ollama'
export { createUnSpeech } from './unspeech'
export { createUnElevenLabs } from './unspeech/elevenlabs'
export { createUnMicrosoft } from './unspeech/microsoft'
