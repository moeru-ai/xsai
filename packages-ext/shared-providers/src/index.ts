export type * from './types/create-provider-options'
export type * from './types/providers'
export * from './utils/create'
export * from './utils/merge'
