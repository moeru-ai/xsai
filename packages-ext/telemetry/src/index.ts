export { generateText } from './utils/generate-text'
export { streamText } from './utils/stream-text'
