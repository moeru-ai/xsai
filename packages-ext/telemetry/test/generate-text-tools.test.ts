import type { ReadableSpan } from '@opentelemetry/sdk-trace-base'

import { InMemorySpanExporter, SimpleSpanProcessor } from '@opentelemetry/sdk-trace-base'
import { NodeTracerProvider } from '@opentelemetry/sdk-trace-node'
import { generateText as aiGenerateText, tool as aiTool, stepCountIs } from 'ai'
import { ollama } from 'ollama-ai-provider-v2'
import { describe, expect, it } from 'vitest'
import { tool } from 'xsai'
import { z } from 'zod/v4'

import { generateText } from '../src'

describe.sequential('generateText with tools', () => {
  const memoryExporter = new InMemorySpanExporter()
  const tracerProvider = new NodeTracerProvider({
    spanProcessors: [new SimpleSpanProcessor(memoryExporter)],
  })
  tracerProvider.register()

  const getAttributes = (span: ReadableSpan) => Object.fromEntries(
    Object.entries(span.attributes)
      .map(([k, v]) => k === 'ai.response.text'
        ? [k, (v as string).replaceAll('\n', ' ').slice(0, 50)]
        : [k, v],
      ),
  )

  it('basic', async () => {
    const add = await tool({
      description: 'Adds two numbers',
      execute: ({ a, b }) => (Number.parseInt(a) + Number.parseInt(b)).toString(),
      name: 'add',
      parameters: z.object({
        a: z.string()
          .describe('First number'),
        b: z.string()
          .describe('Second number'),
      }),
    })

    const { text } = await generateText({
      baseURL: 'http://localhost:11434/v1',
      messages: [{
        content: 'How many times does 114514 plus 1919810 equal? Please try to call the `add` tool to solve the problem.',
        role: 'user',
      }],
      model: 'qwen3:0.6b',
      seed: 114514,
      tools: [add],
    })

    const spans = memoryExporter.getFinishedSpans().map(getAttributes)

    expect(text).toMatchSnapshot()
    expect(spans).toMatchSnapshot()
  })

  it('basic/ai', async () => {
    const add = aiTool({
      description: 'Adds two numbers',
      // @ts-expect-error wrong types
      execute: ({ a, b }) => (Number.parseInt(a) + Number.parseInt(b)).toString(),
      // @ts-expect-error wrong types
      inputSchema: z.object({
        a: z.string()
          .describe('First number'),
        b: z.string()
          .describe('Second number'),
      }),
      name: 'add',
    })

    const { text } = await aiGenerateText({
      experimental_telemetry: { isEnabled: true },
      model: ollama('qwen3:0.6b'),
      prompt: 'How many times does 114514 plus 1919810 equal? Please try to call the `add` tool to solve the problem.',
      seed: 114514,
      stopWhen: stepCountIs(5),
      tools: { add },
    })

    const spans = memoryExporter.getFinishedSpans().slice(4).map(getAttributes)

    expect(text).toMatchSnapshot()
    expect(spans).toMatchSnapshot()
  })

  it('basic/compare', () => {
    const extractAttributeKeys = (span: ReadableSpan) => Object.keys(span.attributes)
      // eslint-disable-next-line sonarjs/no-nested-functions
      .filter(key => ['ai.settings'].every(prefix => !key.startsWith(prefix)))
      .sort((a, b) => a.localeCompare(b))

    const spans = memoryExporter.getFinishedSpans()
    const xsaiSpans = spans.slice(0, 4)
    const aiSpans = spans.slice(4)

    const xsaiIds = xsaiSpans.map(span => span.attributes['ai.operationId'])
    const aiIds = aiSpans.map(span => span.attributes['ai.operationId'])

    expect(xsaiIds).toStrictEqual(aiIds)

    const xsaiKeys = xsaiSpans.map(extractAttributeKeys)
    const aiKeys = aiSpans.map(extractAttributeKeys)

    xsaiKeys.forEach((attributes, i) => {
      expect(attributes).toStrictEqual(aiKeys[i])
    })
  })
})
