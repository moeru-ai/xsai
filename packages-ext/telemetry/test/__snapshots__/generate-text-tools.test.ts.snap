// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateText with tools > basic 1`] = `
"<think>
Okay, so the user asked how many times 114514 plus 1919810 equals. First, I need to calculate that sum. Let me add those two numbers. 114,514 plus 1,919,810. Adding them step by step: 1,919,810 + 100,000 is 2,019,810, then add 14,514 to get 2,034,324. So the total is 2,034,324. The question might have a typo, but based on the numbers provided, the result of the addition is 2,034,324. The user called the add tool, so I'll return that as the answer.
</think>

The sum of 114,514 and 1,919,810 is 2,034,324. 

**Answer:** 2,034,324"
`;

exports[`generateText with tools > basic 2`] = `
[
  {
    "ai.model.id": "qwen3:0.6b",
    "ai.model.provider": "xsai",
    "ai.operationId": "ai.generateText.doGenerate",
    "ai.prompt.messages": "[{"content":"How many times does 114514 plus 1919810 equal? Please try to call the \`add\` tool to solve the problem.","role":"user"}]",
    "ai.prompt.toolChoice": "{"type":"auto"}",
    "ai.prompt.tools": [
      "{"description":"Adds two numbers","inputSchema":{"$schema":"https://json-schema.org/draft/2020-12/schema","type":"object","properties":{"a":{"type":"string"},"b":{"type":"string"}},"required":["a","b"],"additionalProperties":false},"name":"add","type":"function"}",
    ],
    "ai.response.finishReason": "tool_calls",
    "ai.response.id": "c2f28f9c-1641-4082-ae9e-fe5e298131dc",
    "ai.response.model": "qwen3:0.6b",
    "ai.response.providerMetadata": "{}",
    "ai.response.timestamp": "2025-08-30T15:53:28.748Z",
    "ai.response.toolCalls": "[{"args":"{\\"a\\":\\"114514\\",\\"b\\":\\"1919810\\"}","toolCallId":"call_wwh8d495","toolCallType":"function","toolName":"add"}]",
    "ai.usage.completionTokens": 528,
    "ai.usage.promptTokens": 180,
    "gen_ai.request.model": "qwen3:0.6b",
    "gen_ai.response.finish_reasons": [
      "tool_calls",
    ],
    "gen_ai.response.id": "e1476dd3-b5b9-43fc-b49c-12111250ab3e",
    "gen_ai.response.model": "qwen3:0.6b",
    "gen_ai.system": "xsai",
    "gen_ai.usage.input_tokens": 180,
    "gen_ai.usage.output_tokens": 528,
    "operation.name": "ai.generateText.doGenerate",
  },
  {
    "ai.operationId": "ai.toolCall",
    "ai.toolCall.args": "{"a":"114514","b":"1919810"}",
    "ai.toolCall.id": "call_wwh8d495",
    "ai.toolCall.name": "add",
    "ai.toolCall.result": ""2034324"",
    "operation.name": "ai.toolCall",
  },
  {
    "ai.model.id": "qwen3:0.6b",
    "ai.model.provider": "xsai",
    "ai.operationId": "ai.generateText.doGenerate",
    "ai.prompt.messages": "[{"content":"How many times does 114514 plus 1919810 equal? Please try to call the \`add\` tool to solve the problem.","role":"user"},{"role":"assistant","content":"<think>\\nOkay, so the user is asking how many times 114514 plus 1919810 equals. Let me break this down. First, I need to calculate the sum of these two numbers. Let me add them. 114514 plus 1919810. Let's do the math step by step. Adding 114514 and 1,919,810. Well, adding them together, 1,919,810 plus 114,514. Let me add those. 1,919,810 plus 100,000 is 2,019,810, then add the remaining 14,514. 2,019,810 plus 14,514 equals 2,034,324. So the total sum is 2,034,324. But the question is how many times does that equal. Wait, maybe there's a typo or misunderstanding here. The user wrote \\"114514 plus 1919810\\", which I think is the sum. But if the question is phrased as \\"how many times does 114514 plus 1919810 equal?\\" then maybe it's a different interpretation. Perhaps they want to know the sum of the two numbers? Or maybe it's a multiple? Wait, let me check again. The problem says \\"114514 plus 1919810 equals\\". That would mean the sum. Unless there's a specific context where this is a multiplication question, but I don't think so. The user might have made a typo. Alternatively, maybe it's a code or a problem where the sum needs to be input, but given the way the question is phrased, I think the correct answer is the sum, which is 2,034,324. But maybe the user intended to ask for division or something else? Wait, the problem is very straightforward: add the two numbers. So the answer is their sum. Therefore, calling the 'add' function with a and b as the numbers.\\n</think>\\n\\n","tool_calls":[{"id":"call_wwh8d495","index":0,"type":"function","function":{"name":"add","arguments":"{\\"a\\":\\"114514\\",\\"b\\":\\"1919810\\"}"}}]},{"content":"2034324","role":"tool","tool_call_id":"call_wwh8d495"}]",
    "ai.prompt.toolChoice": "{"type":"auto"}",
    "ai.prompt.tools": [
      "{"description":"Adds two numbers","inputSchema":{"$schema":"https://json-schema.org/draft/2020-12/schema","type":"object","properties":{"a":{"type":"string"},"b":{"type":"string"}},"required":["a","b"],"additionalProperties":false},"name":"add","type":"function"}",
    ],
    "ai.response.finishReason": "stop",
    "ai.response.id": "6730d695-e0c4-492a-9830-3a961694cbb3",
    "ai.response.model": "qwen3:0.6b",
    "ai.response.providerMetadata": "{}",
    "ai.response.text": "<think> Okay, so the user asked how many times 114",
    "ai.response.timestamp": "2025-08-30T15:53:31.283Z",
    "ai.usage.completionTokens": 237,
    "ai.usage.promptTokens": 693,
    "gen_ai.request.model": "qwen3:0.6b",
    "gen_ai.response.finish_reasons": [
      "stop",
    ],
    "gen_ai.response.id": "eedae05d-18e9-42a8-be8f-ed28a83d8d7c",
    "gen_ai.response.model": "qwen3:0.6b",
    "gen_ai.system": "xsai",
    "gen_ai.usage.input_tokens": 693,
    "gen_ai.usage.output_tokens": 237,
    "operation.name": "ai.generateText.doGenerate",
  },
  {
    "ai.model.id": "qwen3:0.6b",
    "ai.model.provider": "xsai",
    "ai.operationId": "ai.generateText",
    "ai.prompt": "{"messages":[{"content":"How many times does 114514 plus 1919810 equal? Please try to call the \`add\` tool to solve the problem.","role":"user"}]}",
    "ai.response.finishReason": "stop",
    "ai.response.providerMetadata": "{}",
    "ai.response.text": "<think> Okay, so the user asked how many times 114",
    "ai.usage.completionTokens": 237,
    "ai.usage.promptTokens": 693,
    "operation.name": "ai.generateText",
  },
]
`;

exports[`generateText with tools > basic/ai 1`] = `"20,343,24"`;

exports[`generateText with tools > basic/ai 2`] = `
[
  {
    "ai.model.id": "qwen3:0.6b",
    "ai.model.provider": "ollama.responses",
    "ai.operationId": "ai.generateText.doGenerate",
    "ai.prompt.messages": "[{"role":"user","content":[{"type":"text","text":"How many times does 114514 plus 1919810 equal? Please try to call the \`add\` tool to solve the problem."}]}]",
    "ai.prompt.toolChoice": "{"type":"auto"}",
    "ai.prompt.tools": [
      "{"type":"function","name":"add","description":"Adds two numbers","inputSchema":{"$schema":"http://json-schema.org/draft-07/schema#","type":"object","properties":{"a":{"description":"First number","type":"string"},"b":{"description":"Second number","type":"string"}},"required":["a","b"],"additionalProperties":false}}",
    ],
    "ai.response.finishReason": "stop",
    "ai.response.id": "aitxt-y1axMfbbl765FYwshqrTH9gP",
    "ai.response.model": "qwen3:0.6b",
    "ai.response.providerMetadata": "{"ollama":{}}",
    "ai.response.timestamp": "2025-08-30T15:53:32.544Z",
    "ai.response.toolCalls": "[{"toolCallId":"WdNUDugLi2MWNjL3","toolName":"add","input":"{\\"a\\":\\"114514\\",\\"b\\":\\"1919810\\"}"}]",
    "ai.settings.maxRetries": 2,
    "ai.settings.seed": 114514,
    "ai.usage.completionTokens": 36,
    "ai.usage.promptTokens": 192,
    "gen_ai.request.model": "qwen3:0.6b",
    "gen_ai.response.finish_reasons": [
      "stop",
    ],
    "gen_ai.response.id": "aitxt-y1axMfbbl765FYwshqrTH9gP",
    "gen_ai.response.model": "qwen3:0.6b",
    "gen_ai.system": "ollama.responses",
    "gen_ai.usage.input_tokens": 192,
    "gen_ai.usage.output_tokens": 36,
    "operation.name": "ai.generateText.doGenerate",
  },
  {
    "ai.operationId": "ai.toolCall",
    "ai.toolCall.args": "{"a":"114514","b":"1919810"}",
    "ai.toolCall.id": "WdNUDugLi2MWNjL3",
    "ai.toolCall.name": "add",
    "ai.toolCall.result": ""2034324"",
    "operation.name": "ai.toolCall",
  },
  {
    "ai.model.id": "qwen3:0.6b",
    "ai.model.provider": "ollama.responses",
    "ai.operationId": "ai.generateText.doGenerate",
    "ai.prompt.messages": "[{"role":"user","content":[{"type":"text","text":"How many times does 114514 plus 1919810 equal? Please try to call the \`add\` tool to solve the problem."}]},{"role":"assistant","content":[{"type":"tool-call","toolCallId":"WdNUDugLi2MWNjL3","toolName":"add","input":{"a":"114514","b":"1919810"}}]},{"role":"tool","content":[{"type":"tool-result","toolCallId":"WdNUDugLi2MWNjL3","toolName":"add","output":{"type":"text","value":"2034324"}}]}]",
    "ai.prompt.toolChoice": "{"type":"auto"}",
    "ai.prompt.tools": [
      "{"type":"function","name":"add","description":"Adds two numbers","inputSchema":{"$schema":"http://json-schema.org/draft-07/schema#","type":"object","properties":{"a":{"description":"First number","type":"string"},"b":{"description":"Second number","type":"string"}},"required":["a","b"],"additionalProperties":false}}",
    ],
    "ai.response.finishReason": "stop",
    "ai.response.id": "aitxt-LQFb1FDEQQZUW9MXGGTsLsz8",
    "ai.response.model": "qwen3:0.6b",
    "ai.response.providerMetadata": "{"ollama":{}}",
    "ai.response.text": "20,343,24",
    "ai.response.timestamp": "2025-08-30T15:53:32.750Z",
    "ai.settings.maxRetries": 2,
    "ai.settings.seed": 114514,
    "ai.usage.completionTokens": 10,
    "ai.usage.promptTokens": 245,
    "gen_ai.request.model": "qwen3:0.6b",
    "gen_ai.response.finish_reasons": [
      "stop",
    ],
    "gen_ai.response.id": "aitxt-LQFb1FDEQQZUW9MXGGTsLsz8",
    "gen_ai.response.model": "qwen3:0.6b",
    "gen_ai.system": "ollama.responses",
    "gen_ai.usage.input_tokens": 245,
    "gen_ai.usage.output_tokens": 10,
    "operation.name": "ai.generateText.doGenerate",
  },
  {
    "ai.model.id": "qwen3:0.6b",
    "ai.model.provider": "ollama.responses",
    "ai.operationId": "ai.generateText",
    "ai.prompt": "{"prompt":"How many times does 114514 plus 1919810 equal? Please try to call the \`add\` tool to solve the problem."}",
    "ai.response.finishReason": "stop",
    "ai.response.providerMetadata": "{"ollama":{}}",
    "ai.response.text": "20,343,24",
    "ai.settings.maxRetries": 2,
    "ai.settings.seed": 114514,
    "ai.usage.completionTokens": 10,
    "ai.usage.promptTokens": 245,
    "operation.name": "ai.generateText",
  },
]
`;
